### ISSUE TITLE
Document how to integrate Quarkiverse Amazon DevServices with custom Dev Services

### ISSUE THREAD

**Issue by ryandens** [quarkiverse/quarkus-amazon-services#1527]
I have a quarkus app that uses Amazon S3 that I like to use in quarkus dev mode. I really enjoy and appreciate this extensions ability to start a localstack container and configure my AWS SDK inside the app.

My app has an external dependency on another service that I run locally with Docker. This service also needs access to the same localstack container as my app. I attempted to [write a dev service](https://quarkus.io/guides/extension-writing-dev-service) and can quite easily create another extension that creates a new `DevServicesResultBuildItem` responsible for starting/stopping a docker container.

However, getting access to the localstack container's generated host name on the the network is quite challenging. [This line of code](https://github.com/quarkiverse/quarkus-amazon-services/blob/3f494739f7c7a7eba21a22511e8ca184565a9f9c/common/deployment/src/main/java/io/quarkiverse/amazon/common/deployment/DevServicesLocalStackProcessor.java#L328) is responsible for configuring the network with the hostname `localstack-<random-string>`. 

It would be nice to be able to explicitly depend on the Amazon Dev Service build item in some way shape or form, but I was unable to do that. I can't reliably find that the amazon devservice container starts before my own build item so manually inspecting the docker daemon state and attempting to discover the localstack network/hostname/host port doesn't seem to work either.

Is there some mechanism I'm missing with quarkus extensions?

**Comment by scrocquesel** [comment]
AFAIK, it is currently not possible to orchestrate order of dev services from different extensions. I guess each container should be responsible of waiting for its dependencies to be available to be ready. This should be enough if the dependency graph acyclic and it keeps things simple and isolated.

When not possible, this [discussion](https://github.com/quarkusio/quarkus/discussions/38398) is a good start describing your issue. you could also look at this [DevServices Processor](https://github.com/microcks/microcks-quarkus/blob/main/deployment/src/main/java/io/github/microcks/quarkus/deployment/DevServicesMicrocksProcessor.java) where it implements a consumer of DevServicesLauncherConfigResultBuildItem to ultimately start a container but you will not be able to return dev services properties.

As for, Quarkus Amazon extension, I think that if we add ability to configure a network alias for the container, you should be able to connect to it from your other dev services when using a shared network. The line `ConfigureUtil.configureSharedNetwork` should be rewrited accordingly to a more recent pattern introduced by this [PR](https://github.com/quarkusio/quarkus/pull/39899). Then when shared network is turned on, we can add a predictable network aliases based on the `serviceName` (something like `.withNetworkAliases(serviceName);`) The alias will be `localstack-s3` in dev mode.

Would you like to contribute ?

**Comment by ryandens** [comment]
Hi @scrocquesel - yes i'd be happy to contribute! The network alias makes a lot of sense. One thing I noticed was that the localstack container seems to always be configured to use a shared network today, the `useSharedNetwork` boolean is ignored. Do you want me to update this, potentially a breaking change for consumers of this project?

A way to do this without making this a breaking change would be to update the `useSharedNetwork` parameter to be calculated based on the new API (`DevServicesSharedNetworkBuildItem.isSharedNetworkRequired(devServicesConfig,
                            devServicesSharedNetworkBuildItem)`) as specified in the PR you linked and then to only add the network alias if `useSharedNetwork` is `true`. 

Alternatively, since the container is always configured with a shared network, we could simply always add the network alias like is done in other [quarkus extensions that always use shared networking](https://github.com/search?q=repo%3Aquarkusio%2Fquarkus%20withNetworkAliases&type=code) (keycloak, kerberos, rabbitmq, kafka, etc).

Let me know your thoughts!

**Comment by ryandens** [comment]
By the way, I got my use case to work following the conventions from `microcks` that you linked above! (Creating a build step that consumes `DevServicesLauncherConfigResultBuildItem` and produces an`EmptyBuildItem`. I thought initially this was not an option for me because I was relying on the `DevServicesResultBuildItem.RunningDevService`'s ability to do config overrides to configure my Quarkus application with the container's dynamically assigned host port. Instead, I assigned a static host port so that I don't need to override the config and then was able to access the configured S3 endpoint using `devServicesConfigResult.getConfig().get("quarkus.s3.endpoint-override")`.

**Comment by scrocquesel** [comment]
> Alternatively, since the container is always configured with a shared network, we could simply always add the network alias like is done in other [quarkus extensions that always use shared networking](https://github.com/search?q=repo%3Aquarkusio%2Fquarkus%20withNetworkAliases&type=code) (keycloak, kerberos, rabbitmq, kafka, etc).
> 
> Let me know your thoughts!

I think this is the easiest and conformist solution.

**Comment by ryandens** [comment]
Sounds good, thanks! This will definitely simplify my setup

**Comment by gilday** [comment]
I find myself here, because I am trying to use the localstack dev service with a `QuarkusIntegrationTest` that tests a Quarkus application running as a container. Since the Quarkus app under test is running as a container, it fails to find the localstack service at 127.0.0.1.

If I understand correctly, then the solution discussed here could also allow for this configuration to work.

**Comment by ryandens** [comment]
Hi @scrocquesel  - are you still a maintainer of this repository? I noticed there's been a bit less activity on here as of late, and want to make sure we're reaching the correct person

**Comment by scrocquesel** [comment]
Yes, still I'm a bit busy and work on this on my spare time only. I missed your PR, I'll look into it this week.

@gilday I guess you could force the testcontainer hostname by setting TESTCONTAINERS_HOST_OVERRIDE as described in this page https://java.testcontainers.org/features/configuration/

**Comment by ryandens** [comment]
Thank you! No worries, totally understand. @gilday and I are taking a look to incorporate a more wholistic change to align with other extensions - will update my PR with the changes before this weekend

**Comment by scrocquesel** [comment]
Closing. PR available in 3.4.0+
